### PR TITLE
Remove references to `RefResolver`

### PR DESCRIPTION
**Pull request by jsignell**
**Related Issue(s):**

- #1186 

**Description:**

I am working though the recommendations on https://python-jsonschema.readthedocs.io/en/latest/referencing/ these are the steps as I understand them:

- [x] Replace `RefResolver` with `Registry`
- [x] Create a callable that loads referenced relative schemas
- [x] Create a callable that loads referenced absolute schemas

**PR Checklist:**

- [x] `pre-commit` hooks pass locally
- [x] Tests pass (run `scripts/test`)
- [x] Documentation has been updated to reflect changes, if applicable
- [x] This PR maintains or improves overall codebase code coverage.
- [x] Changes are added to the [CHANGELOG](https://github.com/stac-utils/pystac/blob/main/CHANGELOG.md). See [the docs](https://pystac.readthedocs.io/en/latest/contributing.html#changelog) for information about adding to the changelog.

### CODE REVIEW

**Comment by gadomski** [comment]
> It was never really intended for public consumption.

Yeah, in retrospect `pystac.validation.local_validator` probably should have been `pystac.validation._local_validator`. Maybe we "privatize" the whole module and drop a deprecation warning on the "public" version? E.g. (crib version):

```python
# in pystac/validation/local_validator.py

warnings.warn("This module is deprecated and will be removed in v2")
from ._local_validator import *
__all__ = ["the", "things", "we", "need", "to", "export"]
```

**Review by gadomski — COMMENTED** [review]
Makes sense to me ... I think the only outstanding question is https://github.com/stac-utils/pystac/pull/1215#issuecomment-1718369556.

**Comment by jsignell** [comment]
Hmm the functionality in the old vs new are pretty different. Maybe I should just leave the old `local_validator.py` alone and toss a deprecation warning on it and then create a new `_local_validator.py`.

**Comment by jsignell** [comment]
Ok see what you think about this approach to deprecation. Note that it probably won't have quite as good coverage as main because the deprecated class is untested. I can add some new tests if that is too gross.

File: pystac/validation/local_validator.py [hunk 1]
```diff
@@ -13,6 +15,48 @@
...
+            )
+        },
+    }
+
+
+############################### DEPRECATED #################################
+
 ITEM_SCHEMA_URI = (
     f"https://schemas.stacspec.org/v{VERSION}/item-spec/json-schema/item.json"
 )
```
Comment (gadomski) [inline pystac/validation/local_validator.py:62]: If this constant isn't deprecated, we should move it above the comment. If it is, we should deprecate it with `__getattr__`.

**Review by gadomski — COMMENTED** [review]
Just one small nit then we should be good, thanks!

File: pystac/validation/local_validator.py (hunk 1 above)
Comment (jsignell) [reply pystac/validation/local_validator.py:62]: Yeah I am trying to deprecate those constants. I didn't know about the `__getattr__` trick!

**Review by gadomski — APPROVED** [review]

**Pull request merged**